Add Resource Folder Security Health Analytics Custom Module

vijaykanthm opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
Related to http://b/330538229.
Issue No: https://b.corp.google.com/issues/339718395

Requirement Doc: https://docs.google.com/document/d/1b97vduAO9BPiTTFa-muB9c9ACuDFlAOZW745PrkohXM/edit?tab=t.0#heading=h.cujlsysi3w59

Adding new template to move from SCC API v1 → SCC Management API for post-V2-GA release customers and to support
https://cloud.google.com/security-command-center/docs/reference/security-center-management/rest/v1/folders.locations.securityHealthAnalyticsCustomModules

<!--
Please self-review your PR against the review checklist before creating it: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

Completing the checklist will help speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.

If your PR is still work in progress, please create it in draft mode
-->

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_scc_management_folder_security_health_analytics_custom_module`
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10839-old..auto-pr-10839) ( 13 files changed, 2358 insertions(+), 648 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10839-old..auto-pr-10839) ( 11 files changed, 2454 insertions(+), 754 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-10839-old..auto-pr-10839) ( 1 file changed, 338 insertions(+))


## Errors

`google` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

`google-beta` provider:
- The diff processor failed to build. This is usually due to the downstream provider failing to compile.

SarahFrench commented on mmv1/third_party/terraform/services/securitycentermanagement/resource_scc_management_organization_folder_security_health_analytics_custom_module_test.go:
```diff
@@ -0,0 +1,234 @@
+package securitycenter_test
```
```suggestion
package securitycentermanagement_test
```

vijaykanthm replied:
Addressed

SarahFrench reviewed (changes requested):
👋 Hi! I've included some comments and suggestions in this review, but the main thing is there are currently some build and linting errors being reported in the GitHub action checks:

```
Error: google/provider/provider_mmv1_resources.go:107:2: found packages securitycentermanagement (resource_scc_management_folder_security_health_analytics_custom_module.go) and securitycenter (resource_scc_management_organization_folder_security_health_analytics_custom_module_test.go) in /home/runner/work/magic-modules/magic-modules/google/services/securitycentermanagement
Error: Process completed with exit code 1.
```

```
gofmt needs running on the following files:
./google-beta/services/securitycentermanagement/resource_scc_management_organization_folder_security_health_analytics_custom_module_test.go
```

Please:
- fmt the Go code in services/securitycentermanagement/resource_scc_management_organization_folder_security_health_analytics_custom_module_test.go
- Check the package names in .../services/securitycentermanagement/

SarahFrench commented on mmv1/products/securitycentermanagement/FolderSecurityHealthAnalyticsCustomModule.yaml:
```diff
@@ -0,0 +1,235 @@
+      characters or underscores only.
+  - !ruby/object:Api::Type::Enum
+    name: 'enablementState'
+    required: false
```
When these fields are unset they default to false, and the norm is to not include these fields unless we set them to true. Could you please remove this line, and others like it?

vijaykanthm replied:
Thank you for the info. Addressed.

vijaykanthm commented on mmv1/products/securitycentermanagement/FolderSecurityHealthAnalyticsCustomModule.yaml:
```diff
@@ -0,0 +1,235 @@
+    required: false
+    description: |
+      The enablement state of the custom module.
+    values:
+      - :ENABLED
+      - :DISABLED
```
The v1 version doesn't have the `INHERITED` value, so I haven't mentioned it. Do you think it is good to have?

SarahFrench replied:
Oh, my mistake. I saw the value `INHERITED` in [the docs here](https://cloud.google.com/security-command-center/docs/reference/security-center-management/rest/v1/folders.locations.securityHealthAnalyticsCustomModules) but I have less context on this work than you do (I am a HashiCorp employee)

vijaykanthm commented on mmv1/products/securitycentermanagement/FolderSecurityHealthAnalyticsCustomModule.yaml:
```diff
@@ -0,0 +1,235 @@
+  - !ruby/object:Api::Type::String
+    name: 'displayName'
+    immutable: true
+    required: false
```
Addressed

vijaykanthm commented on mmv1/products/securitycentermanagement/FolderSecurityHealthAnalyticsCustomModule.yaml:
```diff
@@ -0,0 +1,235 @@
+      specifies the organization or folder from which the custom module is inherited.
+  - !ruby/object:Api::Type::NestedObject
+    name: 'customConfig'
+    required: false
```
Addressed

vijaykanthm commented:
Hi @SarahFrench , can you please have a look at the PR, I have addressed the comments and made some changes, can you approve the step to trigger the next steps in the CI. Thanks.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10839-old..auto-pr-10839) ( 5 files changed, 1668 insertions(+), 2 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10839-old..auto-pr-10839) ( 5 files changed, 1668 insertions(+), 2 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-10839-old..auto-pr-10839) ( 1 file changed, 338 insertions(+))



## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_scc_management_folder_security_health_analytics_custom_module` (0 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_scc_management_folder_security_health_analytics_custom_module" "primary" {
  custom_config {
    custom_output {
      properties {
        name = # value needed
        value_expression {
          description = # value needed
          expression  = # value needed
          location    = # value needed
          title       = # value needed
        }
      }
    }
    description = # value needed
    predicate {
      description = # value needed
      expression  = # value needed
      location    = # value needed
      title       = # value needed
    }
    recommendation = # value needed
    resource_selector {
      resource_types = # value needed
    }
    severity = # value needed
  }
  display_name     = # value needed
  enablement_state = # value needed
  folder           = # value needed
  location         = # value needed
}


```

modular-magician commented:
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$
`TestAccSecurityCenterManagementOrganizationFolderSecurityHealthAnalyticsCustomModule`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10839/artifacts/4035f3df-6089-42da-b839-2b6d17395ea7/build-log/recording_build/TestAccSecurityCenterManagementOrganizationFolderSecurityHealthAnalyticsCustomModule_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10839/artifacts/4035f3df-6089-42da-b839-2b6d17395ea7/recording/TestAccSecurityCenterManagementOrganizationFolderSecurityHealthAnalyticsCustomModule.log)]


$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10839/artifacts/4035f3df-6089-42da-b839-2b6d17395ea7/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-10839/artifacts/4035f3df-6089-42da-b839-2b6d17395ea7/recording) for each test

SarahFrench commented on mmv1/products/securitycentermanagement/FolderSecurityHealthAnalyticsCustomModule.yaml:
```diff
@@ -0,0 +1,232 @@
+# limitations under the License.
+
+--- !ruby/object:Api::Resource
+name: 'FolderSecurityHealthAnalyticsCustomModule'
```
The name FolderSecurityHealthAnalyticsCustomModule results in the new resource having this name:
- google_scc_management_folder_security_health_analytics_custom_module

But in your acceptance tests you are referring to the resource as:
- google_scc_management_organization_folder_security_health_analytics_custom_module

Could you please either update the name here (and the file name) _or_ update the acceptance test to use the correct resource name?

vijaykanthm replied:
Thank you for the insight. Addressed.

SarahFrench replied:
Thanks!

SarahFrench reviewed (changes requested):
Left a comment about a mismatch between the resource name in the tests and in the provider - I think fixing that would address the lack of test coverage [that automation is raising here](https://github.com/GoogleCloudPlatform/magic-modules/pull/10839#issuecomment-2152732158)

SarahFrench commented on mmv1/products/securitycentermanagement/product.yaml:
```diff
@@ -0,0 +1,23 @@
+--- !ruby/object:Api::Product
+name: SecurityCenterManagement
+display_name: Security Command Center Management (SCC)
+legacy_name: scc_management
```
To add to my previous comment, the resource name is also impacted by this value

vijaykanthm replied:
I have changed the resource name in the config, test case and example files. So I think the `legacy name` shouldn't affect  the resource now, is that right?

SarahFrench replied:
I believe legacy name controls this section of the resource type's name: google_**scc_management**\_folder_security_health_analytics_custom_module.

Everything's ok with the legacy name currently and you don't need to change it; I highlighted legacy_name in a comment in case you wanted the resource type name to start with google_**scc_management_organization**\_ , because then we'd need to decide whether to achieve that by updating `legacy_name` in product.yaml or `name` in FolderSecurityHealthAnalyticsCustomModule.yaml.

Given the change in https://github.com/GoogleCloudPlatform/magic-modules/pull/10839/commits/d0b8ac7a1bf685d8292378759202743caa8383c3#diff-908bcb8f9291d65c4bb97b2f734cea4d65ef3b6310e12a580d5f3196cb696e9a this isn't necessary

vijaykanthm replied:
Cool, thanks.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10839-old..auto-pr-10839) ( 5 files changed, 1668 insertions(+), 2 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10839-old..auto-pr-10839) ( 5 files changed, 1668 insertions(+), 2 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-10839-old..auto-pr-10839) ( 1 file changed, 338 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: `2`
Passed tests: `1`
Skipped tests: `0`
Affected tests: `1`

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>securitycentermanagement</li></ul></blockquote></details>


#### Action taken
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccSecurityCenterManagementFolderSecurityHealthAnalyticsCustomModule </blockquote></details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$
`TestAccSecurityCenterManagementFolderSecurityHealthAnalyticsCustomModule`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10839/artifacts/d165ea63-8418-42e2-b38e-5673478268e4/recording/TestAccSecurityCenterManagementFolderSecurityHealthAnalyticsCustomModule.log)]


$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$

---
$\textcolor{green}{\textsf{All tests passed!}}$
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10839/artifacts/d165ea63-8418-42e2-b38e-5673478268e4/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-10839/artifacts/d165ea63-8418-42e2-b38e-5673478268e4/recording) for each test

SarahFrench commented on mmv1/products/securitycentermanagement/FolderSecurityHealthAnalyticsCustomModule.yaml:
```diff
@@ -0,0 +1,232 @@
+          help investigators understand the detected issue. The text must be enclosed in quotation marks.
+      - !ruby/object:Api::Type::String
+        name: 'recommendation'
+        required: true
```
Is recommendation required? In the API docs it's listed as Optional on this page: https://cloud.google.com/security-command-center/docs/reference/security-center-management/rest/v1/CustomConfig

vijaykanthm replied:
You are right, these are Optional, I have removed the field. Thanks.

SarahFrench commented on mmv1/products/securitycentermanagement/FolderSecurityHealthAnalyticsCustomModule.yaml:
```diff
@@ -0,0 +1,232 @@
+            item_type: Api::Type::String
+      - !ruby/object:Api::Type::Enum
+        name: 'severity'
+        required: true
```
Is severity required? In the API docs it's listed as Optional on this page: https://cloud.google.com/security-command-center/docs/reference/security-center-management/rest/v1/CustomConfig

vijaykanthm replied:
Addressed.

SarahFrench commented on mmv1/products/securitycentermanagement/FolderSecurityHealthAnalyticsCustomModule.yaml:
```diff
@@ -0,0 +1,232 @@
+                        file name and a position in the file.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'resourceSelector'
+        required: true
```
Is resourceSelector required? In the API docs it's listed as Optional on this page: https://cloud.google.com/security-command-center/docs/reference/security-center-management/rest/v1/CustomConfig

vijaykanthm replied:
Addressed.

SarahFrench commented on mmv1/products/securitycentermanagement/FolderSecurityHealthAnalyticsCustomModule.yaml:
```diff
@@ -0,0 +1,232 @@
+    properties:
+      - !ruby/object:Api::Type::NestedObject
+        name: 'predicate'
+        required: true
```
Is predicate required? In the API docs it's listed as Optional on this page: https://cloud.google.com/security-command-center/docs/reference/security-center-management/rest/v1/CustomConfig

vijaykanthm replied:
Addressed.

SarahFrench reviewed (commented):
I'm glad the tests passed! Before approving I took a last look at the resource's fields to check for any differences between the YAML file and the docs. I've made comments where I found some.

I don't have access to the same internal information as you do, so I appreciate the API docs could be incorrect. If the fields I commented on _are_ indeed required please just reply to the comments and mark resolved.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10839-old..auto-pr-10839) ( 5 files changed, 1668 insertions(+), 2 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10839-old..auto-pr-10839) ( 5 files changed, 1668 insertions(+), 2 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-10839-old..auto-pr-10839) ( 1 file changed, 338 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: `2`
Passed tests: `2`
Skipped tests: `0`
Affected tests: `0`

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>securitycentermanagement</li></ul></blockquote></details>

$\textcolor{green}{\textsf{All tests passed!}}$
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10839/artifacts/e4798113-637d-4dcb-8e41-fe556b4a29ba/build-log/replaying_test.log)

vijaykanthm commented:
> Left a comment about a mismatch between the resource name in the tests and in the provider - I think fixing that would address the lack of test coverage [that automation is raising here](https://github.com/GoogleCloudPlatform/magic-modules/pull/10839#issuecomment-2152732158)

Addressed

vijaykanthm commented:
Hi @SarahFrench , I have made the changes as per your suggestions, everything looks good now and all the tests are also passed. Can you please have a look and approve if everything is good. Thanks.

SarahFrench commented:
I'll merge in a few hours time, waiting on CI
